packet.go

```diff
@@ -0,0 +1,217 @@
+
+package gonat
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"syscall"
+
+	"github.com/getlantern/errors"
+)
+
+// TCPFlags are the different flags supported in the TCP header
+const (
+	TCPFlagRST = 0x04
+)
+
+var (
+	networkByteOrder = binary.BigEndian
+)
+
+func parseIPPacket(raw []byte) (*IPPacket, error) {
+	ipVersion := uint8(raw[0]) >> 4
+	if ipVersion != 4 {
+		return nil, errors.New("Unsupported ip protocol version: %v", ipVersion)
+	}
+
+	pkt := &IPPacket{Raw: raw, IPVersion: ipVersion}
+	return pkt.parseV4()
+}
+
+type IPPacket struct {
+	Raw       []byte
+	IPVersion uint8
+	IPProto   uint8
+	SrcAddr   *net.IPAddr
+	DstAddr   *net.IPAddr
+	Header    []byte
+	Payload   []byte
+}
+
+func (pkt *IPPacket) parseV4() (*IPPacket, error) {
+	ihl := uint8(pkt.Raw[0]) & 0x0F
+	length := networkByteOrder.Uint16(pkt.Raw[2:4])
+	if length < 20 {
+		return pkt, errors.New("Invalid (too small) IP length (%d < 20)", length)
+	} else if ihl < 5 {
+		return pkt, errors.New("Invalid (too small) IP header length (%d < 5)", ihl)
+	} else if int(ihl*4) > int(length) {
+		return pkt, errors.New("Invalid IP header length > IP length (%d > %d)", ihl, length)
+	} else if int(ihl)*4 > len(pkt.Raw) {
+		return pkt, errors.New("Not all IP header bytes available")
+	}
+
+	pkt.Header = pkt.Raw[:ihl*4]
+	pkt.Payload = pkt.Raw[ihl*4:]
+	pkt.IPProto = uint8(pkt.Header[9])
+	pkt.SrcAddr = &net.IPAddr{IP: net.IP(pkt.Header[12:16])}
+	pkt.DstAddr = &net.IPAddr{IP: net.IP(pkt.Header[16:20])}
+
+	return pkt, nil
+}
+
+// HasTCPFlag returns true if the packet is a TCP packet that has the given flag set.
+func (pkt *IPPacket) HasTCPFlag(flag uint8) bool {
+	return pkt.IPProto == syscall.IPPROTO_TCP && pkt.Payload[13]&flag != 0
+}
+
+func (pkt *IPPacket) SetSource(addr Addr) {
+	ip := addr.IP()
+	copy(pkt.Header[12:16], ip)
+	networkByteOrder.PutUint16(pkt.Payload[0:2], addr.Port)
+	pkt.SrcAddr = &net.IPAddr{IP: ip}
+}
+
+func (pkt *IPPacket) SetDest(addr Addr) {
+	ip := addr.IP()
+	copy(pkt.Header[16:20], ip)
+	networkByteOrder.PutUint16(pkt.Payload[2:4], addr.Port)
+	pkt.DstAddr = &net.IPAddr{IP: ip}
+}
+
+func (pkt *IPPacket) ipChecksum() uint16 {
+	return networkByteOrder.Uint16(pkt.Header[10:])
+}
+
+func (pkt *IPPacket) recalcIPChecksum() {
+	// Clear checksum bytes
+	pkt.Header[10] = 0
+	pkt.Header[11] = 0
+
+	// Compute checksum
+	var csum uint32
+	for i := 0; i < len(pkt.Header); i += 2 {
+		csum += uint32(pkt.Header[i]) << 8
+		csum += uint32(pkt.Header[i+1])
+	}
+	for {
+		// Break when sum is less or equals to 0xFFFF
+		if csum <= 65535 {
+			break
+		}
+		// Add carry to the sum
+		csum = (csum >> 16) + uint32(uint16(csum))
+	}
+	// Flip all the bits
+	networkByteOrder.PutUint16(pkt.Header[10:], ^uint16(csum))
+}
+
+func (pkt *IPPacket) tcpChecksum() uint16 {
+	return networkByteOrder.Uint16(pkt.Payload[16:])
+}
+
+func (pkt *IPPacket) udpChecksum() uint16 {
+	return networkByteOrder.Uint16(pkt.Payload[6:])
+}
```

Comment:
Why do we have this, tcpChecksum, and ipChecksum?  Just for the tests?

Reply:
Yeah, and for logging when I was debugging some stuff.